### PR TITLE
Add GNU compiler flags to help build with GNU v10 and above

### PR DESCRIPTION
**Pull request by ndkeen**
Specifically, adding -fallow-argument-mismatch -fallow-invalid-boz for any GNU build using version 10 or higher.

Fixes https://github.com/E3SM-Project/E3SM/issues/3954

[bfb]

### CODE REVIEW

File: cime_config/machines/cmake_macros/gnu.cmake [hunk 1]
```diff
@@ -40,6 +40,9 @@ endif()
 if (NOT DEBUG)
   string(APPEND FFLAGS " -O")
 endif()
+if (CMAKE_Fortran_COMPILER_VERSION VERSION_GREATER_EQUAL 10)
+   string(APPEND FFLAGS " -fallow-argument-mismatch -fallow-invalid-boz")
```
Comment (jayeshkrishna) [inline cime_config/machines/cmake_macros/gnu.cmake:44]: We also found that we needed the "-fallow-argument-mismatch" flag for building SCORPIO with gnu 11

**Review by jayeshkrishna — APPROVED** [review]

**Comment by ndkeen** [comment]
Note it seems like these are work-around flags and there may be other ways to address.

The `-fallow-argument-mismatch` flag is used to make these warnings instead of errors:
```
   22 |     call MPI_BCAST(vec,lsize,MPI_INTEGER,0,comm,ierr)
      |                   1
......
   31 |     call MPI_BCAST(vec,lsize,MPI_LOGICAL,0,comm,ierr)
      |                   2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/LOGICAL(4)).
```

The `-fallow-invalid-boz` flag is actually only needed for one line in the code.
In `components/eam/src/control/cam_history.F90`, there is this line:
`integer, parameter :: gen_hash_key_offset = z'000053db'`

Which looks to be used to init a hash before adding values. Not familiar

File: cime_config/machines/cmake_macros/gnu.cmake (hunk 1 above)
Comment (ndkeen) [reply cime_config/machines/cmake_macros/gnu.cmake:44]: I suspect you need it for GNU v10 as well

**Comment by jayeshkrishna** [comment]
Are the changes BFB (BFB flag is missing in the PR desc, and the PR labels)?

**Comment by ndkeen** [comment]
I added a commit to the branch to change how we issue the `-fallow-invalid-boz` flag. Now it's only used for 2 specific files via Depends.gnu*.  Re-tested e3sm_developer.

**Comment by ndkeen** [comment]
Closing this in favor of https://github.com/E3SM-Project/E3SM/pull/4822

**Pull request closed without merging**